### PR TITLE
Sfu/buffer stats for telemetry

### PR DESCRIPTION
**Pull request by frostbyte73**

### CODE REVIEW

File: pkg/sfu/streamallocator.go [hunk 1]
```diff
@@ -939,7 +939,7 @@ func (s *StreamAllocator) isTimeToBoost() bool {
 }
 
 func (s *StreamAllocator) resetBoost() {
-	s.lastBoostTime = time.UnixMilli(0)
+	s.lastBoostTime = time.Unix(0, 0)
```
Comment (boks1971) [inline pkg/sfu/streamallocator.go:942]: I think DZ pushed a change for this already.

File: pkg/telemetry/events.go [hunk 2]
```diff
@@ -8,37 +8,44 @@ import (
...
+func (t *TelemetryService) ParticipantJoined(ctx context.Context, room *livekit.Room, participant *livekit.ParticipantInfo) {
+	t.Lock()
+	if t.workers[participant.Sid] == nil {
+		t.workers[participant.Sid] = NewStatsWorker(func(diff *buffer.Stats) {
+			t.HandleIncomingRTP(participant.Sid, participant.Identity, diff)
+		})
+	}
+	t.Unlock()
+
+	prometheus.AddParticipant()
```
Comment (boks1971) [inline pkg/telemetry/events.go:46]: Maybe add only when adding a NewStatsWorker?

File: pkg/telemetry/events.go [hunk 3]
```diff
@@ -47,10 +54,17 @@ func (s *TelemetryService) ParticipantJoined(ctx context.Context, room *livekit.
...
-	s.pool.Submit(prometheus.SubParticipant)
+func (t *TelemetryService) ParticipantLeft(ctx context.Context, room *livekit.Room, participant *livekit.ParticipantInfo) {
+	t.Lock()
+	if w := t.workers[participant.Sid]; w != nil {
+		w.Close()
+		delete(t.workers, participant.Sid)
+	}
+	t.Unlock()
+
+	prometheus.SubParticipant()
```
Comment (boks1971) [inline pkg/telemetry/events.go:65]: Maybe check that a participant was really deleted before calling subtract?

File: pkg/telemetry/events.go [hunk 4]
```diff
@@ -59,42 +73,44 @@ func (s *TelemetryService) ParticipantLeft(ctx context.Context, room *livekit.Ro
...
-		prometheus.AddPublishedTrack(track.Type.String())
-	})
+func (t *TelemetryService) TrackPublished(participantID, identity string, track *livekit.TrackInfo, buff *buffer.Buffer) {
+	t.RLock()
+	if w := t.workers[participantID]; w != nil {
+		w.AddBuffer(buff)
+	}
+	t.RUnlock()
+
+	prometheus.AddPublishedTrack(track.Type.String())
```
Comment (boks1971) [inline pkg/telemetry/events.go:83]: Add here only when adding buffer?

File: pkg/telemetry/events.go [hunk 5]
```diff
@@ -59,42 +73,44 @@ func (s *TelemetryService) ParticipantLeft(ctx context.Context, room *livekit.Ro
...
-		prometheus.SubPublishedTrack(track.Type.String())
-	})
+func (t *TelemetryService) TrackUnpublished(participantID, identity string, track *livekit.TrackInfo, ssrc uint32) {
+	t.RLock()
+	if w := t.workers[participantID]; w != nil {
+		w.RemoveBuffer(ssrc)
+	}
+	t.RUnlock()
+
+	prometheus.SubPublishedTrack(track.Type.String())
```
Comment (boks1971) [inline pkg/telemetry/events.go:95]: same as above

File: pkg/telemetry/statsworker.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+		stats := buff.GetStats()
+		total.PacketCount += stats.PacketCount
+		total.TotalByte += stats.TotalByte
+		total.LastExpected += stats.LastExpected
+		total.LastReceived += stats.LastReceived
+		total.LostRate += stats.LostRate
+		total.Jitter += stats.Jitter
+	}
+	s.RUnlock()
+	total.LostRate /= float32(numTracks)
```
Comment (boks1971) [inline pkg/telemetry/statsworker.go:68]: We should not average loss rate and jitter. A loss of 1 packet in a 10 packet stream will skew this. Overall loss should be calculated across all tracks.

For jitter, I am thinking taking the max is more useful.

**Review by boks1971 — APPROVED** [review]
looks good @frostbyte73 . Some small comments!

File: pkg/telemetry/events.go (hunk 3 above)
Comment (frostbyte73) [reply pkg/telemetry/events.go:65]: it should always be deleted, but it should be easy to spot either way if counts are off

**Pull request merged**